gui/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,615 @@
+        Default method for calculating vector lengths
+        Implemented ONLY for image-like vector data
+        :param newlen: new length
+        :return:
+        '''
+
+        if self._data_type == 'projection':
+            return None
+        elif self._data_type == 'matrix':
+            self._length = newlen
+            self._vectors = self._check_vector_type(self._original_data)
+
+    @property
+    def color(self) -> str:
+        """Color, ColorArray: color of the body of the marker
+        """
+        return self._color
+
+    @color.setter
+    def color(self, color: str) -> None:
+        self._color = color
+        self._refresh()
+
+    @property
+    def connector(self):
+        """
+        line connector.  One of _connector.types = "strip", "segments"
+        or can receive an ndarray.
+        :return: connector parameter
+        """
+        return self._connector
+
+    @connector.setter
+    def connector(self, connector_type: Union[str, np.ndarray]):
+        self._connector = connector_type
+        self._refresh()
+
+    @property
+    def mode(self):
+        """None, str: Interactive mode
+        """
+        return self._mode
+
+    @mode.setter
+    def mode(self, mode):
+        if mode == self.mode:
+            return
+        if mode == 'pan/zoom':
+            self.cursor = 'standard'
+            self.interactive = True
+            self.help = ''
+            self.status = mode
+            self._mode = mode
+        else:
+            raise ValueError("Mode not recongnized")
+
+        self.events.mode(mode=mode)
+
+    # =========================== Napari Layer ABC methods ===================
+    @property
+    def data(self) -> np.ndarray:
+        """
+
+        :return: coordinates of line vertices
+        """
+        return self.vectors
+
+    @data.setter
+    def data(self, data: np.ndarray) -> None:
+        """
+        :param data:
+        :return:
+        """
+
+        self.vectors = data
+
+    def _get_shape(self):
+        if len(self.vectors) == 0:
+            return np.ones(self.vectors.shape, dtype=int)
+        else:
+            return np.max(self.vectors, axis=0) + 1
+
+    def _refresh(self):
+        """Fully refresh the underlying visual.
+        """
+        self._need_display_update = True
+        self._update()
+
+    def _update(self):
+        """Update the underlying visual.
+        """
+        if self._need_display_update:
+            self._need_display_update = False
+
+            self._set_view_slice(self.viewer.dims.indices)
+
+        if self._need_visual_update:
+            self._need_visual_update = False
+            self._node.update()
+
+    def _set_view_slice(self, indices):
+        """Sets the view given the indices to slice with.
+
+        Parameters
+        ----------
+        indices : sequence of int or slice
+            Indices to slice with.
+        """
+        in_slice_vectors, matches = self._slice_vectors(indices)
```

Comment:
If you see my comment bellow about deleting `_slice_vectors`, I would just remove this line and have
`in_slice_vectors = self.vectors`

Reply:
Doing this will mean that they layer will not crash if you try and overlay the 2D vectors layer on a 3D or 4D image, instead the 2D vectors will always appear regardless of what slice of the image you are looking at, which I think is fine for now, and we can work on an nD vectors layer after the new dims refactoring